packages/plugin-ext/src/hosted/node/hosted-plugin-process.ts

```diff
@@ -93,34 +92,42 @@ export class HostedPluginProcess implements ServerPluginRunner {
         const cp = this.childProcess;
         this.childProcess = undefined;
 
-        const emitter = new Emitter();
+        const waitForTerminated = new Deferred<void>();
         cp.on('message', message => {
-            emitter.fire(JSON.parse(message));
-        });
-        const rpc = new RPCProtocolImpl({
-            onMessage: emitter.event,
-            send: (m: {}) => {
-                if (cp.send) {
-                    cp.send(JSON.stringify(m));
-                }
+            const msg = JSON.parse(message);
+            if ('type' in msg && msg.type === MessageType.Terminated) {
+                waitForTerminated.resolve();
             }
         });
-        const hostedPluginManager = rpc.getProxy(MAIN_RPC_CONTEXT.HOSTED_PLUGIN_MANAGER_EXT);
-        hostedPluginManager.$stop().then(() => {
-            emitter.dispose();
-            this.killProcessTree(cp.pid);
-        });
+        cp.send(JSON.stringify({ type: MessageType.Terminate }));
+        // give 10 seconds to terminate gracefully and then kill the process
+        await Promise.race([
+            waitForTerminated,
+            new Promise(resolve => setTimeout(resolve, 10 * 1000))
+        ]);
+        this.killProcessTree(cp.pid);
     }
 
     private killProcessTree(parentPid: number): void {
-        psTree(parentPid, (err: Error, childProcesses: Array<psTree.PS>) => {
-            childProcesses.forEach((p: psTree.PS) => {
-                process.kill(parseInt(p.PID));
-            });
-            process.kill(parentPid);
+        psTree(parentPid, (_, childProcesses) => {
+            childProcesses.forEach(childProcess =>
+                this.killProcess(parseInt(childProcess.PID))
+            );
+            this.killProcess(parentPid);
         });
     }
 
+    protected killProcess(pid: number): void {
+        try {
+            process.kill(pid);
+        } catch (e) {
+            if (e && 'code' in e && e.code === 'ESRCH') {
+                return;
+            }
+            console.error(`[${pid}] failed to kill`, e);
```

Comment:
sorry, it was there for logger instead of console.error

Reply:
ok, will use here